Add "open cohort in notebook" tool [SATURN-1168]

melissachang opened the pull request:
@jeromechadel approved this flow. Lou said I could use this Jupyter icon. (I ran `npm run optimize-image-svgs`.)

FYI @seanhorgan @wnojopra @mbookman 

Notebook tool button:

![1](https://user-images.githubusercontent.com/10929390/66441759-10ce7e80-e9ed-11e9-9ecf-7e48f74af30a.png)

Tooltip is "Create a Python 2 or 3 notebook with this notebook". I wanted to clarify that this works for Python and not R.

After you click the button, notebook creation dialog:

![1](https://user-images.githubusercontent.com/10929390/66246231-9f7a8d00-e6c7-11e9-9c6d-e60c5afc9df2.png)

Then you're taken to notebook. The [notebook is here](https://app.terra.bio/#workspaces/fc-product-demo/melchang%20test/notebooks/launch/cohort-notebook.ipynb) if anyone wants to see it. It includes a visualization.

![1](https://user-images.githubusercontent.com/10929390/66246238-b4572080-e6c7-11e9-9762-89f02781e411.png)

Tested:
- Notebook name dialog:
  - Error shown if you type in existing name
  - Clicking cancel and outside dialog work as expected
- Notebook
  - Works with Py2 and Py3
  - cohorts from 1000 Genomes, Framingham, Baseline, AMP PD
  - By default the [fapi library](https://github.com/broadinstitute/fiss) talks to prod (api.firecloud.org). I didn't see an easy way to change this, so I wasn't able to run the notebook in local Terra. I did test the notebook in prod Terra.
  - Workspace has BigQuery tables from multiple datasets. Correct BigQuery tables are used
- Normal way of creating notebook still works

Will merge after #1805

seanhorgan commented:
Thanks @melissachang, great to see this come together. I like the language note in the tooltip. Should we also clarify that this action isn't just an open but a create? I'd be interested in Jerome's thoughts here especially in the case where a user has already created a notebook from this cohort and wants to open it again. This could open a door to cohort provenance both in a workspace and beyond.

Can you share that test notebook with me (my @google address)?

seanhorgan commented:
One other question: is there a way to monitor usage? I'd like to be able to understand adoption, e.g. how many unique notebooks were created from a cohort in the last month, how many unique people created a notebook from a cohort in the last 3 months?

melissachang commented:
@seanhorgan - You should be able to see [notebook](https://app.terra.bio/#workspaces/fc-product-demo/melchang%20test/notebooks/launch/cohort-notebook.ipynb) now.

FYI due to [IA-1402](https://broadworkbench.atlassian.net/browse/IA-1402) you have to run a cell, restart kernel, then run all cells. Once [IA-1402](https://broadworkbench.atlassian.net/browse/IA-1402) is fixed, then you can simply run all cells.

That's a good point about `create` vs `open`. I changed tooltip to "Open cohort in a new Python 2 or 3 notebook". FYI Brett suggested only supporting `create` to begin with, so I don't have worry about supporting locking/syncing. I wanted to get something working; then it can be iterated upon.

Regarding monitoring -- I haven't been following Terra client-side logging. Once there is a client-side logging system in place, it won't be hard to add this use case.

melissachang committed:
Add "open cohort in notebook" tool

change help article name

new tooltip

fix running cell twice to see plot

onDismiss/onSuccess

flow

review fixes

add provenance cell

melissachang commented:
This is ready to be reviewed. I rebased with #1805 and retested.

jeromechadel commented:
Open with a new notebook / open with an existing notebook.
There’s no need of a new pattern here.
Jerome

> On Oct 6, 2019, at 11:03 AM, Sean Horgan <notifications@github.com> wrote:
> 
> Thanks @melissachang <https://github.com/melissachang>, great to see this come together. I like the language note in the tooltip. Should we also clarify that this action isn't just an open but a create? I'd be interested in Jerome's thoughts here especially in the case where a user has already created a notebook from this cohort and wants to open it again. This could open a door to cohort provenance both in a workspace and beyond.
> 
> Can you share that test notebook with me (my @google <https://github.com/google> address)?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=ALLHY3RTDK7PLU6U6N2I7T3QNH44PA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAOMERY#issuecomment-538755655>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALLHY3VUQPFAMHC6CGY4ELDQNH44PANCNFSM4I5VO3PA>.
>

melissachang commented:
@jeromechadel -- Brett suggested to **initially** only support `create`, so I don't have to worry about locking/syncing. Can this PR support only `create`, then someone (not me) can add support for `open` if they want?

jeromechadel commented:
Yes, you are correct. I was just answering Sean. :-)
Jerome

> On Oct 8, 2019, at 6:42 PM, melissachang <notifications@github.com> wrote:
> 
> @jeromechadel <https://github.com/jeromechadel> -- Brett suggested to initially only support create, so I don't have to worry about locking/syncing. Can this PR support only create, then someone (not me) can add support for open if they want?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=ALLHY3S4LT3EQ2362XFQG5LQNUEEXA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV3BVY#issuecomment-539734231>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ALLHY3QGOYAY3R3WMPIQ7CDQNUEEXANCNFSM4I5VO3PA>.
>

seanhorgan commented:
Great. Sounds like we are aligned that this action should be referred to as
'create'.

On Tue, Oct 8, 2019 at 6:44 PM jeromechadel <notifications@github.com>
wrote:

> Yes, you are correct. I was just answering Sean. :-)
> Jerome
>
> > On Oct 8, 2019, at 6:42 PM, melissachang <notifications@github.com>
> wrote:
> >
> > @jeromechadel <https://github.com/jeromechadel> -- Brett suggested to
> initially only support create, so I don't have to worry about
> locking/syncing. Can this PR support only create, then someone (not me) can
> add support for open if they want?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <
> https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=ALLHY3S4LT3EQ2362XFQG5LQNUEEXA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV3BVY#issuecomment-539734231>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/ALLHY3QGOYAY3R3WMPIQ7CDQNUEEXANCNFSM4I5VO3PA
> >.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=AADIWOOQ4WLNZ35OF77TTQDQNUELLA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV3FQQ#issuecomment-539734722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADIWOKIGA4DEHPDU73J3MTQNUELLANCNFSM4I5VO3PA>
> .
>

melissachang commented:
Got it. Tooltip is now "Create a Python 2 or 3 notebook with this cohort".

On Tue, Oct 8, 2019 at 4:48 PM Sean Horgan <notifications@github.com> wrote:

> Great. Sounds like we are aligned that this action should be referred to as
> 'create'.
>
> On Tue, Oct 8, 2019 at 6:44 PM jeromechadel <notifications@github.com>
> wrote:
>
> > Yes, you are correct. I was just answering Sean. :-)
> > Jerome
> >
> > > On Oct 8, 2019, at 6:42 PM, melissachang <notifications@github.com>
> > wrote:
> > >
> > > @jeromechadel <https://github.com/jeromechadel> -- Brett suggested to
> > initially only support create, so I don't have to worry about
> > locking/syncing. Can this PR support only create, then someone (not me)
> can
> > add support for open if they want?
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub <
> >
> https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=ALLHY3S4LT3EQ2362XFQG5LQNUEEXA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV3BVY#issuecomment-539734231
> >,
> > or mute the thread <
> >
> https://github.com/notifications/unsubscribe-auth/ALLHY3QGOYAY3R3WMPIQ7CDQNUEEXANCNFSM4I5VO3PA
> > >.
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=AADIWOOQ4WLNZ35OF77TTQDQNUELLA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV3FQQ#issuecomment-539734722
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AADIWOKIGA4DEHPDU73J3MTQNUELLANCNFSM4I5VO3PA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataBiosphere/terra-ui/pull/1881?email_source=notifications&email_token=ACTMJ3VE3JFRS25EO36KVYDQNUL4VA5CNFSM4I5VO3PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAV7EZY#issuecomment-539751015>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACTMJ3XYLCYTEXJYUSJME63QNUL4VANCNFSM4I5VO3PA>
> .
>

zarsky-broad commented on src/components/notebook-utils.js:
```diff
@@ -78,7 +79,7 @@ export const NotebookCreator = class NotebookCreator extends Component {
   render() {
     const { notebookName, notebookKernel, creating, nameTouched } = this.state
-    const { reloadList, onDismiss, namespace, bucketName, existingNames } = this.props
+    const { reloadList, onDismissNotebookCreated, onDismissNotebookNotCreated, namespace, bucketName, existingNames } = this.props
```
This seems like `onDismiss` vs `onSuccess`.

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -360,14 +364,37 @@ const ReferenceDataContent = ({ workspace: { workspace: { namespace, attributes
 
-const ToolDrawer = withModalDrawer()(({
-  workspace, workspace: { workspace: { workspaceId } }, onDismiss, onIgvSuccess, entityMetadata, entityKey, selectedEntities
+const ToolDrawer = withModalDrawer()(requesterPaysWrapper({
```
This should use `_.flow` for composition.

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -360,14 +364,37 @@ const ReferenceDataContent = ({ workspace: { workspace: { namespace, attributes
+  workspace, workspace: { workspace: { workspaceId } }, onDismiss, onIgvSuccess,
+  onRequesterPaysError, entityMetadata, entityKey, selectedEntities
 }) => {
+  const { workspace: { bucketName, name: wsName, namespace } } = workspace
```
Why not add these to the destructure above?

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -360,14 +364,37 @@ const ReferenceDataContent = ({ workspace: { workspace: { namespace, attributes
+    setNotebookNames(notebooks.map(notebook => notebook.name.slice(10, -6)))
+  })
+  Utils.useOnMount(() => {
+    loadNotebookNames()
```
If this is only used here, our convention is to move the definition inside of the `useOnMount` as well, for clarity.

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -360,14 +364,37 @@ const ReferenceDataContent = ({ workspace: { workspace: { namespace, attributes
-  const dataExplorerButtonEnabled = isCohort && entitiesCount === 1 && _.values(selectedEntities)[0].attributes.data_explorer_url
+  const dataExplorerButtonEnabled = isCohort && entitiesCount === 1 && _.values(selectedEntities)[0].attributes.data_explorer_url !== undefined
   const dataExplorerUrl = dataExplorerButtonEnabled ? `${_.values(selectedEntities)[0].attributes.data_explorer_url}&wid=${workspaceId}` : undefined
+  const notebookButtonEnabled = isCohort && entitiesCount === 1
```
If you put this before `dataExplorerButtonEnabled`, it can build off of this.

melissachang replied:
I think it's cleaner and less bug-prone to be explicit

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -382,6 +409,24 @@ const ToolDrawer = withModalDrawer()(({
+        bucketName, namespace,
+        existingNames: notebookNames,
+        onDismissNotebookCreated: async notebookName => {
+          console.assert(isCohort && entitiesCount === 1, '1 cohort should be selected')
```
Is this necessary? We appear to be guarding for it above.

melissachang replied:
Not with this PR. I thought I would add it to prevent future bugs, since this code is complicated. Removed.

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -382,6 +409,24 @@ const ToolDrawer = withModalDrawer()(({
+          const cohortName = _.values(selectedEntities)[0].name
+          const contents = cohortNotebookTemplate.replace('MY_COHORT', cohortName)
+          await Buckets.notebook(namespace, bucketName, notebookName).create(JSON.parse(contents))
+          Nav.goToPath('workspace-notebook-launch', { namespace, name: wsName, notebookName: notebookName + '.ipynb' })
```
```suggestion
          Nav.goToPath('workspace-notebook-launch', { namespace, name: wsName, notebookName: `${notebookName}.ipynb` })
```

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -382,6 +409,24 @@ const ToolDrawer = withModalDrawer()(({
+        onDismissNotebookNotCreated: () => {
+          setToolMode(undefined)
+        },
+        reloadList: () => {}
```
```suggestion
        reloadList: _.noop
```

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -382,6 +409,24 @@ const ToolDrawer = withModalDrawer()(({
+          await Buckets.notebook(namespace, bucketName, notebookName).create(JSON.parse(contents))
+          Nav.goToPath('workspace-notebook-launch', { namespace, name: wsName, notebookName: notebookName + '.ipynb' })
+        },
+        onDismissNotebookNotCreated: () => {
```
```suggestion
        onDismissNotebookNotCreated: setToolMode
```

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -425,6 +470,23 @@ const ToolDrawer = withModalDrawer()(({
+                [isCohort && entitiesCount > 1, () => 'Select exactly one cohort to open in notebook'],
+                [!isCohort, () => 'Only cohorts can be opened with notebooks'],
+                [notebookButtonEnabled, () => 'Create a Python 2 or 3 notebook with this cohort'],
+                () => undefined
```
Unneeded, undefined is the implicit default.

melissachang replied:
Done

zarsky-broad reviewed (commented):
Instead of having `cohort-notebook` as an ipynb file, consider defining it as a function that takes the name and just returns JSON, and putting that in a js file. That would remove the need for the new loader, and also allow you to avoid doing string replacement where you're creating it.

melissachang reviewed (commented):
Notebooks are typically represented as .ipynb files. I think it's cleaner as-is. Is that ok?

Typical workflow:
- Make changes to notebook in Terra. Then download `.ipynb`, copy to this repo

zarsky-broad reviewed (commented):
Adding in a new loader still feels like a lot to me, especially since it's to load a file with a custom extension that we're just loading the text out of. It also feels quite brittle to be leaning on that string substitution for the cohort name. I'd rather see one of the following two options:

```js
export const cohortNotebook = cohortName => ({
  'cells': [
   {
    'cell_type': 'code',
    'execution_count': null,
    'metadata': {},
    'outputs': [],
    'source': [
     `COHORT = '${cohortName}'`
    ]
   },
...
})
```
or
```js
export const cohortNotebook = cohortName => `
{
  "cells": [
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
    "source": [
     "COHORT = '${cohortName}'"
    ]
   },
...
`
```

At the end, this gets stringified, so it may make the most sense just to treat it as a big string and sub in the one value; either way, I think string substitution on a file that we're adding with a custom extension so we can ignore the extension and parse it as text is an unnecessary complication. The benefit of the object/non-string version is that we can make sure we don't break the structure of the file.

melissachang committed:
merge

melissachang committed:
put notebook template in js

melissachang committed:
fix typo

melissachang committed:
rm template

melissachang commented:
Done, put notebook template in js

zarsky-broad reviewed (approved):
Clear to merge.

TOL: It occurs to me that the double-escaping of newlines is something we wouldn't have to worry about if we treated the template as JSON instead of a big string, but that's pretty minor.

melissachang commented:
Thanks. I'll merge after [WORKSPACE_NAME bug](https://broadinstitute.zendesk.com/agent/tickets/3776) is fixed (should be today or next couple days). Notebooks won't work with that bug.
